src/odemis/driver/orsay.py

```diff
@@ -2614,3 +2643,224 @@ def terminate(self):
+    objective lens a is a constant coefficient stored in MD_CALIB (should equal 0.18e6 V/m) and d is the relative change
+    in focus distance in meter.
+    """
+
+    def __init__(self, name, role, parent, rng, **kwargs):
+        """
+        Initialise Focus. Raises AttributeError exception if there is no _fib_beam sibling.
+        :param (tuple (float, float)) rng: the range of the z axis
+        """
+        axes_def = {"z": model.Axis(unit="m", range=rng)}
+        super().__init__(name, role, parent=parent, axes=axes_def, **kwargs)
+
+        self.position = model.VigilantAttribute({"z": 0.0}, readonly=True, unit="m")
+        self.parent._fib_beam.objectiveVoltage.subscribe(self._updatePosition)
+
+        self.baseLensVoltage = 0.0  # V1, the objective lens voltage corresponding to a focus distance of 0
+        # Changes during runtime
+
+        # Event that get's set when the position VA changes value while performing a move using _doMoveAbs
+        self._position_changed = threading.Event()
+        self._position_changed.clear()
+
+        self._executor = CancellableThreadPoolExecutor(max_workers=1)
+
+        self.updateMetadata({})
+
+    def updateMetadata(self, md):
+        md[model.MD_CALIB] = 0.18e6
+        super().updateMetadata(md)
+
+    def _updatePosition(self, value=None):
+        """
+        Calculates the current focus distance as d = (value - baseLensVoltage) / MD_CALIB
+        :param (float or None) value: the current value of the objective lens voltage. If value is None, this voltage
+        will be read from the FIBBeam sibling.
+        """
+        if value is None:
+            value = self.parent._fib_beam.objectiveVoltage.value
+        new_d = (value - self.baseLensVoltage) / self._metadata[model.MD_CALIB]
+        self.position._set_value({"z": new_d}, force_write=True)
+
+    def _doMoveRel(self, shift):
+        """
+        Calculate the new position of the focus and pass the result to _doMoveAbs
+        :param (dict {"z": value}) shift: value contains the desired change in focus position in meter.
+        """
+        shift["z"] += self.position.value["z"]
+        self._doMoveAbs(shift)
+
+    def _position_changed_event_handler(self, value):
+        self._position_changed.set()
+
+    def _doMoveAbs(self, pos):
+        """
+        Calculated the new voltage as V = baseLensVoltage + MD_CALIB * (current_position + delta).
+        Blocking until the new position is reached or it times out.
+        :param (dict {"z": value}) pos: value contains the desired new focus position in meter.
+        """
+        self.position.subscribe(self._position_changed_event_handler)
+
+        new_voltage = self.baseLensVoltage + self._metadata[model.MD_CALIB] * pos["z"]
+        self.parent._fib_beam.objectiveVoltage.value = new_voltage
+        logging.debug(f"Focus position set to {pos['z']}, setting objective voltage to {new_voltage}")
+
+        start = time.time()
```

Comment:
To simplify a tiny bit, instead of storing `start` you can store `endt = time.time() + FOCUS_CHANGE_TIMEOUT `.